### DISCUSSION TITLE
Basic mask and color swap

### DISCUSSION THREAD

**Discussion by RoYoMi** [ImageMagick/ImageMagick#4371] [Help] [answered]
I have this starting image which shows a series of UI buttons on screen. The number of the current active button is slightly larger and the circle is shown in light blue.  Any of the six buttons could be active and when activated they move left or right slightly.

![image](https://user-images.githubusercontent.com/8366040/137500955-54bb6607-8aa2-4642-b4d4-cd7b4a88fd25.png)

I can take the image and mask it down to just the active blue circle using the following command:
> convert /Users/royomi/Desktop/ButtonActive.png -fuzz 15% -fill black +opaque "#25C4EA" -opaque "#231f20" -threshold 0% -colorspace RGB -background black -flatten +matte /Users/royomi/Desktop/ButtonActiveMasked.png

![image](https://user-images.githubusercontent.com/8366040/137501545-b605db50-7f75-4d79-88a2-ce9b45a5d4d9.png)

It looks ok, but it's not great.

I have two problems I can't quite solve.

1. How do I keep the black inside the blue circle?
2. How do I fill the outside of the circle with a known color?

I think this would be done with some basic masking and a floodfill but I'm new to image processing and this all is a bit confusing to me.

Ideally I'd be left with a black and white image with the black from inside the blue circle remains black, and the rest of the image is turned white, or the otherway around.

This is being done on a Mac with OSX 11.5.2
Convert version: ImageMagick 7.1.0-10 Q16-HDRI x86_64 2021-10-10

**Comment by snibgo** [comment]
As you use IM v7, I suggest you use "magick", not "convert".

You have "-colorspace RGB", which converts to linear RGB. Why do you do this?

"+matte" is deprecated. Use "-alpha off" instead.

Your output is blue and transparent. The command you showed should make an image of black and white pixels only. What command made your image?

I'm not sure what you want. Perhaps you want to isolate that blue button. If so, then you need a mask that black where the button is (including the "2") and white elsewhere. Or the opposite colours, of course.

**Comment by RoYoMi** [comment]
Why use `-colorspace RGB`, without this it throws an error. Apparently the image is converted to a grayscale and won't save as a PNG. `convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG`
I found the file can be saved without `-colorspace RGB` if it's saved as a JPG instead of a PNG.

> Your output is blue and transparent. The command you showed should make an image of black and white pixels only. What command made your image?

Your observation highlights how little I know of about image processing. When I open the image in Apple Preview it appears like you said black and white. To bring the image into this forum I used copy & paste. When it got pasted into the forum here it then looked blue and white. I thought it looked odd too so I tried it a couple times. Eventually I thought it was just an artifact of this processing, or preview wasn't showing me the correct thing.

Here is what the two look like on my system, using the command

> magick /Users/royomi/Desktop/ButtonActive.png -fuzz 15% -fill black +opaque "#25C4EA" -opaque "#231f20" -threshold 0% -colorspace RGB -background black -flatten -alpha off /Users/royomi/Desktop/ButtonActiveMasked.png 

![image](https://user-images.githubusercontent.com/8366040/137532506-4b353fbe-c8f7-4d21-950c-f42e83cff03e.png)


What I'm looking to do is process image so that all the black background in the second image is filled with white leaving just the black digit. 

When I try the following command with a floodfill statement (syntax from [this page](https://imagemagick.org/script/command-line-options.php#floodfill))it gives the same result. It feels like I'm doing something wrong.

> magick /Users/royomi/Desktop/ButtonActive.png -fuzz 15% -fill black +opaque "#25C4EA" -opaque "#231f20" -threshold 0% -background black -floodfill +1+1 black -flatten -alpha off /Users/royomi/Desktop/ButtonActiveMasked.jpg

![image](https://user-images.githubusercontent.com/8366040/137555973-51595021-5574-4615-b886-2a04c8594a07.png)

**Comment by fmw42** [comment, marked as answer]
This is one way to do what you want, by flood filling with white from the center. Seems to me that you have -colorspace and -flattening when you don't needed. Also -colorspace RGB is linear and not the usual sRGB non-linear colorspace.

```
magick ButtonActive.png \
-alpha off \
-fuzz 15% -fill black +opaque "#25C4EA" -opaque "#231f20" \
-threshold 0% \
-fill white -draw "color %[fx:w/2],%[fx:h/2] floodfill" -alpha off \
ButtonActiveMasked.png

```
if you want it to be colorspace gray

or 

```
magick ButtonActive.png \
-alpha off \
-fuzz 15% -fill black +opaque "#25C4EA" -opaque "#231f20" \
-threshold 0% \
-fill white -draw "color %[fx:w/2],%[fx:h/2] floodfill" -alpha off \
PNG8:ButtonActiveMasked.png

```
if you want it to be 8-bit color and colorspace sRGB

or

```
magick ButtonActive.png \
-alpha off \
-fuzz 15% -fill black +opaque "#25C4EA" -opaque "#231f20" \
-threshold 0% \
-fill white -draw "color %[fx:w/2],%[fx:h/2] floodfill" -alpha off \
PNG24:ButtonActiveMasked.png

```
if you want it to be 24-bit color and colorspace sRGB

![ButtonActiveMasked](https://user-images.githubusercontent.com/13458218/137557951-41b3a94c-6254-4c20-b912-7e1e5fa2832d.png)

**Reply by RoYoMi** [reply]
Thank you that's exactly what I needed.
